src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
-public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public abstract class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
 
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+        size = 0;
     }
 
-    @Override
-    public void add(T value, int index) {
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > elements.length) {
+            int newCapacity = elements.length * 3 / 2 + 1;
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
 
+    private void rangeCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public boolean add(T value) {
+        ensureCapacity(size + 1);
+        elements[size++] = value;
+        return true;
+    }
+
+    public void add(T value, int index) {
+        rangeCheck(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
+    }
 
+    public void addAll(List<T> list) {
+        ensureCapacity(size + list.size());
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) elements[index];
     }
 
-    @Override
     public void set(T value, int index) {
-
+        rangeCheck(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T oldValue = (T) elements[index];
+        int numMoved = size - index - 1;
+        if (numMoved > 0) {
+            System.arraycopy(elements, index + 1, elements, index, numMoved);
+        }
+        elements[--size] = null;
+        return oldValue;
     }
 
-    @Override
-    public T remove(T element) {
-        return null;
+    public boolean remove(Object element) {
+        int index = indexOf(element);
+        if (index >= 0) {
+            remove(index);
+            return true;
+        } else {
+            throw new NoSuchElementException("Element not found");
```

Comment:
The `remove(Object element)` method should not throw a `NoSuchElementException` if the element is not found. Instead, it should return `false`, as this is the typical behavior for a `List` implementation.